### PR TITLE
Add option to perform fast synchronisation runs in File Provider sync engine

### LINKED ISSUE #5692: Improve initial sync of macOS File Provider module for VFS
### ⚠️ Before submitting, please verify the following: ⚠️

- [X] This is a **bug**, not a question or a configuration issue.
- [X] This issue is **not** already reported on Github (I've searched it).
- [X] Nextcloud Server and Desktop Client are **up to date**. See [Server Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) and [Desktop Releases](https://nextcloud.com/install/#install-clients) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/)

### Bug description

The initial full enumeration of the subject server done by the file provider module is extremely slow, leading to a large amount of time passing before the user is able to interact with files in the relevant domain

### Steps to reproduce

1. Set up account with file provider module enabled
2. Wait for enumeration of server files to begin
3. Wait for enumeration of server files to end
4. Wait some more

### Expected behavior

Delay between a file provider enumeration beginning and it being responsive should be minimal

### Which files are affected by this bug

All files affected

### Operating system

Mac OS

### Which version of the operating system you are running.

macOS 13.0

### Package

Other

### Nextcloud Server version

23.0.1

### Nextcloud Desktop Client version

Master

### Is this bug present after an update or on a fresh install?

Fresh desktop client install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### Are you using an external user-backend?

- [ ] Default internal user-backend
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Nextcloud Server logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by claucambra**
Closes #5692 

Screenshot 2024-02-19 at 19 45 49

### CODE REVIEW

**Comment by roberix** [comment]
Hi. 

Is this going to be a mac only feature or will this be also available on the Windows client?

Regards 

rob

**Comment by claucambra** [comment]
> Hi.
> 
> Is this going to be a mac only feature or will this be also available on the Windows client?
> 
> Regards
> 
> rob

Hi, this is a macOS-only feature

File: src/gui/macOS/fileproviderxpc_mac.mm [hunk 1]
```diff
@@ -158,4 +158,11 @@
     return std::optional<std::pair<bool, bool>>{{receivedFastEnumerationEnabled, receivedFastEnumerationSet}};
 }
 
+void FileProviderXPC::setFastEnumerationEnabledForExtension(const QString &extensionAccountId, bool enabled) const
+{
+    qCInfo(lcFileProviderXPC) << "Setting fast enumeration for extension" << extensionAccountId << "to" << enabled;
+    const auto service = (NSObject<ClientCommunicationProtocol> *)_clientCommServices.value(extensionAccountId);
+    [service setFastEnumerationEnabled:enabled];
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/macOS/fileproviderxpc_mac.mm:165]: The XPC service for client to file provider extension communication is being called, in this case to toggle the file provider extension's fast enumeration setting

File: src/gui/macOS/ui/FileProviderFastEnumerationSettings.qml [hunk 2]
```diff
@@ -39,4 +39,21 @@ Column {
...
+            color: Style.infoBoxBackgroundColor
+            border.width: Style.infoBoxBorderWidth
+            border.color: Style.infoBoxBorderColor
+            radius: Style.slightlyRoundedButtonRadius
+        }
+        width: parent.width
+        padding: Style.smallSpacing
+        text: qsTr("Fast sync will only sync changes in files and folders within folders that have been explored. " +
+                    "This can significantly increase responsiveness on initial configuration of virtual files. " +
+                    "However, it will cause redundant downloads of files moved to an unexplored folder. ")
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/macOS/ui/FileProviderFastEnumerationSettings.qml:55]: I'm not sure how to explain this more concisely without losing clarity; I think there is a risk of this setting not being clearly understood

**Pull request merged**